### PR TITLE
Crate serialization major refactor & standarization

### PR DESCRIPTION
**Pull request by CPerezz**
This PR introduces the following major changes:
- Remove support for Serde #353
- Add `dusk-bytes::Serializable` for type of fixed size #352
- Update API naming convention used for variable-size types #354
- Remove code duplication #437
- Add `dusk-bytes` functions instead of custom ones for internal fields #437

Resolves: #352, #354, #437
See also: #353

### CODE REVIEW

**Comment by CPerezz** [comment]
The problem with the PR resides on the fact that when we create proofs on a normal way, the Polynomials stored inside of the `widgets::ProverKey`'s are not of the same lenght. And that causes a serialization error while at the same time is not logical.

So we will need to address this here too.

File: src/fft/domain.rs [hunk 1]
```diff
@@ -14,35 +14,71 @@
...
+    pub(crate) generator_inv: BlsScalar,
 }
 
-impl fmt::Debug for EvaluationDomain {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Multiplicative subgroup of size {}", self.size)
+impl Serializable<{ u64::SIZE + u32::SIZE + 5 * BlsScalar::SIZE }> for EvaluationDomain {
+    type Error = dusk_bytes::Error;
+
+    #[allow(unused_must_use)]
```
Comment (vlopes11) [inline src/fft/domain.rs:46]: What is being unused here?

File: src/lib.rs [hunk 2]
```diff
@@ -52,9 +52,6 @@
 #![deny(broken_intra_doc_links)]
```
Comment (vlopes11) [inline src/lib.rs:52]: Compiler warning:

lint `broken_intra_doc_links` has been renamed to `rustdoc::broken_intra_doc_links`

File: src/proof_system/linearisation_poly.rs [hunk 3]
```diff
@@ -57,54 +55,55 @@ pub struct ProofEvaluations {
...
-        bytes[224..256].copy_from_slice(&self.q_arith_eval.to_bytes()[..]);
-        bytes[256..288].copy_from_slice(&self.q_c_eval.to_bytes()[..]);
-        bytes[288..320].copy_from_slice(&self.q_l_eval.to_bytes()[..]);
-        bytes[320..352].copy_from_slice(&self.q_r_eval.to_bytes()[..]);
-        bytes[352..384].copy_from_slice(&self.left_sigma_eval.to_bytes()[..]);
-        bytes[384..416].copy_from_slice(&self.right_sigma_eval.to_bytes()[..]);
-        bytes[416..448].copy_from_slice(&self.out_sigma_eval.to_bytes()[..]);
-        bytes[448..480].copy_from_slice(&self.lin_poly_eval.to_bytes()[..]);
-        bytes[480..512].copy_from_slice(&self.perm_eval.to_bytes()[..]);
+    #[allow(unused_must_use)]
```
Comment (vlopes11) [inline src/proof_system/linearisation_poly.rs:61]: Same as above; nothing apparently being unused

File: src/proof_system/proof.rs [hunk 4]
```diff
@@ -89,46 +82,48 @@ impl<S: Store> Canon<S> for Proof {
...
-        bytes[384..432].copy_from_slice(&self.t_4_comm.0.to_bytes()[..]);
-        bytes[432..480].copy_from_slice(&self.w_z_comm.0.to_bytes()[..]);
-        bytes[480..528].copy_from_slice(&self.w_zw_comm.0.to_bytes()[..]);
-        bytes[528..PROOF_SIZE].copy_from_slice(&self.evaluations.to_bytes()[..]);
-
-        bytes
+impl Serializable<{ 11 * Commitment::SIZE + ProofEvaluations::SIZE }> for Proof {
+    type Error = Error;
+
+    #[allow(unused_must_use)]
```
Comment (vlopes11) [inline src/proof_system/proof.rs:88]: Same as above

File: src/proof_system/widget/arithmetic/verifierkey.rs [hunk 5]
```diff
@@ -19,6 +20,47 @@ pub struct VerifierKey {
     pub q_arith: Commitment,
 }
 
+impl Serializable<{ 7 * Commitment::SIZE }> for VerifierKey {
+    type Error = dusk_bytes::Error;
+
+    #[allow(unused_must_use)]
```
Comment (vlopes11) [inline src/proof_system/widget/arithmetic/verifierkey.rs:26]: Same as above

File: src/proof_system/widget/mod.rs [hunk 6]
```diff
@@ -60,83 +37,85 @@ pub struct VerifierKey {
     pub permutation: permutation::VerifierKey,
 }
 
-impl_serde!(ProverKey);
-impl_serde!(VerifierKey);
+impl Serializable<{ 15 * Commitment::SIZE + u64::SIZE }> for VerifierKey {
+    type Error = dusk_bytes::Error;
+
+    #[allow(unused_must_use)]
```
Comment (vlopes11) [inline src/proof_system/widget/mod.rs:43]: Same as above

**Review by vlopes11 — APPROVED** [review]
LGTM, small nits to check.

Just be sure we have test cases for all the to/from bytes operations

File: src/fft/domain.rs (hunk 1 above)
Comment (CPerezz) [reply src/fft/domain.rs:46]: The results from `writer.write()`.

File: src/lib.rs (hunk 2 above)
Comment (CPerezz) [reply src/lib.rs:52]: Will address on a sepparate PR. Thanks!

**Review by ZER0 — CHANGES_REQUESTED** [review]
There are few changes to be made. Some are mandatory IMVHO, some other not.

I believe one of the major issue (and API breaking, but we're already breaking the API) is the introduction of `dusk-bytes`.

Since now there are structures that implements `dusk-bytes` and other don't, it's imperative to have the proper naming convention, otherwise it might be very messy.

To be clear:
1. If a type implement a method that have the same signature of dusk-bytes, and perform the same work, it should have the same name, even if it's not implementing `dusk-bytes`. (e.g. `from_slice`)
2. If a type implement a method that have different signature of dusk-bytes, it shouldn't be called in the same way (e.g. `from_bytes` to read slices, and `to_bytes` to ouput vectors)

File: src/commitment_scheme/kzg10/key.rs [hunk 7]
```diff
@@ -40,7 +25,14 @@ pub struct CommitKey {
...
-    /// Will consume twice the bytes of `into_bytes`
+    /// This operation is designed to store the raw representation of the contents
+    /// of the CommitKey. Therefore, the size of the bytes outputed by this function
+    /// is expected to be the double than the one that `CommitKey::to_bytes()`.
+    ///
+    /// # Note
+    /// This function should be used when we want to serialize the CommitKey allowing
+    /// a really fast deserialization later.
+    /// This functions output should not be used by the regular `CommitKey::from_bytes()` fn.
     pub fn to_raw_bytes(&self) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply src/commitment_scheme/kzg10/key.rs:36]: `to_raw_var_bytes` was the agreed new name.

File: src/commitment_scheme/kzg10/key.rs [hunk 8]
```diff
@@ -40,7 +25,14 @@ pub struct CommitKey {
...
         let mut bytes = Vec::with_capacity(8 + self.powers_of_g.len() * G1Affine::RAW_SIZE);
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply src/commitment_scheme/kzg10/key.rs:37]: I missed that. The general idea is to use `u64::SIZE` in these cases. Amended

File: src/fft/domain.rs (hunk 1 above)
Comment (vlopes11) [reply src/fft/domain.rs:46]: Maybe we should put a comment there explaining the situation?

**Review by ZER0 — APPROVED** [review]

**Pull request merged**